// rollup-plugin-banner2/dist/rollup-plugin-banner2.es.js
/**
 * rollup-plugin-banner2@1.3.1
 */
import MagicString from 'magic-string';

const identity = (str) => str;

const docBlock = (str) => {
  const lines = str.split('\n');
  const wrapped = ['/**', ...lines.map((l) => ` * ${l}`), ' */\n'].join('\n');

  return wrapped
};

const docBlockAndGap = (str) => {
  return `${docBlock(str)}\n`
};

const formatters = { docBlock, docBlockAndGap, identity };

const defaultOptions = {
  sourcemap: true,
  formatter: 'identity',
};
function banner2(resolveBanner, userOptions) {
  const opts = { ...defaultOptions, ...(userOptions || {}) };

  return {
    name: 'banner2',
    async renderChunk(code, chunk, options) {
      const banner = await resolveBanner(chunk, options);
      if (!banner) return { code, map: null }

      const formatterByOption = formatters[opts.formatter];
      const formattedBanner = formatterByOption(String(banner));

      if (!opts.sourcemap) return formattedBanner + code

      const magicString = new MagicString(code);
      magicString.prepend(formattedBanner);

      return {
        code: magicString.toString(),
        map: magicString.generateMap({ hires: true }),
      }
    },
  }
}

export default banner2;
